core/config/config_helper.hpp

```diff
@@ -0,0 +1,220 @@
+
+/**
+ * get_factory_vector will gives a vector of factory by calling get_factory.
+ */
+template <typename T>
+inline std::vector<deferred_factory_parameter<T>> get_factory_vector(
+    const pnode& config, const registry& context, type_descriptor td);
+
+
+/**
+ * get_value gets the corresponding type value from config.
+ *
+ * This is specialization for integral type
+ */
+template <typename IndexType>
+inline
+    typename std::enable_if<std::is_integral<IndexType>::value, IndexType>::type
+    get_value(const pnode& config);
+
+/**
+ * get_value gets the corresponding type value from config.
+ *
+ * This is specialization for floating point type
+ */
+template <typename ValueType>
+inline typename std::enable_if<std::is_floating_point<ValueType>::value,
+                               ValueType>::type
+get_value(const pnode& config);
+
+/**
+ * get_value gets the corresponding type value from config.
+ *
+ * This is specialization for complex type
+ */
+template <typename ValueType>
+inline typename std::enable_if<gko::is_complex_s<ValueType>::value,
+                               ValueType>::type
+get_value(const pnode& config);
+
+
+// type_string providing the mapping from type to string.
+template <typename T>
+struct type_string {};
+
+#define TYPE_STRING_OVERLOAD(_type, _str)         \
+    template <>                                   \
+    struct type_string<_type> {                   \
+        static std::string str() { return _str; } \
+    }
+
+TYPE_STRING_OVERLOAD(void, "void");
+TYPE_STRING_OVERLOAD(double, "double");
+TYPE_STRING_OVERLOAD(float, "float");
+TYPE_STRING_OVERLOAD(std::complex<double>, "complex<double>");
+TYPE_STRING_OVERLOAD(std::complex<float>, "complex<float>");
+TYPE_STRING_OVERLOAD(int32, "int");
+TYPE_STRING_OVERLOAD(int64, "int64");
+
+
+template <typename T>
+inline std::shared_ptr<T> get_pointer(const pnode& config,
+                                      const registry& context,
+                                      type_descriptor td)
+{
+    std::shared_ptr<T> ptr;
+    using T_non_const = std::remove_const_t<T>;
+    ptr = context.search_data<T_non_const>(config.get_string());
+    assert(ptr.get() != nullptr);
+    return std::move(ptr);
+}
+
+
+template <typename T>
+inline std::vector<deferred_factory_parameter<T>> get_factory_vector(
+    const pnode& config, const registry& context, type_descriptor td)
+{
+    std::vector<deferred_factory_parameter<T>> res;
+    // for loop in config
+    if (config.get_tag() == pnode::tag_t::array) {
+        for (const auto& it : config.get_array()) {
+            res.push_back(get_factory<T>(it, context, td));
+        }
+    } else {
+        // only one config can be passed without array
+        res.push_back(get_factory<T>(config, context, td));
+    }
+
+    return res;
+}
+
+
+template <typename IndexType>
+inline
+    typename std::enable_if<std::is_integral<IndexType>::value, IndexType>::type
+    get_value(const pnode& config)
```

Comment:
I know we separate declaration and definition in our public headers, but for this implementation header, I would prefer to combine it again. If I'm looking at a private header, I'm more interested in the code, than the docs.